src/metabase/util/ui_logic.clj

```diff
@@ -119,21 +119,78 @@
   from the visualization settings if the column is specified, or from our default column logic"
   [card result]
-  (if-let [user-specified-rowfn (y-axis-rowfn card result)]
-    user-specified-rowfn
-    (when-let [default-col-index (default-goal-column-index card result)]
+  (let [user-specified-rowfn (y-axis-rowfn card result)
+        default-col-index (default-goal-column-index card result)
+        progress-value-col (when (= :progress (:display card))
+                             (get-in card [:visualization_settings :progress.value]))
+        progress-col-index (when progress-value-col
+                             (column-name->index progress-value-col result))
+        first-numeric-col-index (when (and (= :progress (:display card))
+                                           (not progress-col-index))
+                                  (first (keep-indexed (fn [idx col]
+                                                         (when (isa? (:base_type col) :type/Number)
+                                                           idx))
+                                                       (:cols result))))]
+    (cond
+      user-specified-rowfn
+      user-specified-rowfn
+
+      progress-col-index
       (fn [row]
-        (nth row default-col-index)))))
+        (nth row progress-col-index))
+
+      first-numeric-col-index
+      (fn [row]
+        (nth row first-numeric-col-index))
+
+      default-col-index
+      (fn [row]
+        (nth row default-col-index))
+
+      :else nil)))
+
+(defn- extract-goal-value-from-column
+  "Extracts goal value from a column reference, similar to frontend getGoalValue"
+  [goal-setting columns rows]
+  (when (string? goal-setting)
+    (let [column-index (->> columns
+                            (map-indexed vector)
+                            (filter #(= goal-setting (:name (second %))))
+                            (first)
+                            (first))]
+      (when (and column-index rows (seq rows))
+        (let [raw-value (nth (first rows) column-index nil)]
+          (cond
+            (nil? raw-value) 0
+            (= "Infinity" raw-value) ##Inf
+            (number? raw-value) raw-value
+            :else 0))))))
 
 (defn find-goal-value
-  "The goal value can come from a progress goal or a graph goal_value depending on it's type"
+  "The goal value can come from a progress goal or a graph goal_value depending on it's type.
+  For progress charts, the goal can be either a number or a column reference.
+  Matches the frontend behavior: invalid goals fallback to default value (0)."
   [result]
   (case (get-in result [:card :display])
 
     (:area :bar :line)
     (get-in result [:card :visualization_settings :graph.goal_value])
 
     :progress
-    (get-in result [:card :visualization_settings :progress.goal])
+    (let [goal-setting (get-in result [:card :visualization_settings :progress.goal])
```

Comment:
There isn't a way for a numeric goal to be serialized to json as a string right?